kubelet: add internal metric for the first pod with network latency

aojea opened the pull request:
The first pod with network latency impact user workloads, however,
it is difficuly to understand where is the problem of this latency,
since it depends on the CNI plugin to be ready at the moment of the
pod creation.

Add a new internal metric in the kubelet that allow developers and cluster
administrator to understand the source of the latency problems on
node startups.

kubelet_first_network_pod_start_sli_duration_seconds


/kind documentation
```release-note
Add a new internal metric in the kubelet that allow developers to understand the source of the latency problems on node startups.

kubelet_first_network_pod_start_sli_duration_seconds
```

/sig instrumentation
/sig node
/sig network

aojea commented:
/assign @SergeyKanzhelev @dchen1107 @logicalhan @wojtek-t @dgrisonnet 
/cc @Ramkumar-K

aojea commented:
/cc @ruiwen-zhao @qiutongs

logicalhan commented on pkg/kubelet/metrics/metrics.go:
```diff
@@ -211,6 +213,30 @@ var (
+	// excluding the time for image pulling. This metric should reflect the "Pod startup latency SLI" definition and measure the time
+	// for the CNI plugin to be able to provide the network capabilities to the Pod.
+	// ref: https://github.com/kubernetes/community/blob/master/sig-scalability/slos/pod_startup_latency.md
+	FirstNetworkPodStartSLIDuration = metrics.NewGauge(
+		&metrics.GaugeOpts{
+			Subsystem:      KubeletSubsystem,
+			Name:           FirstNetworkPodStartSLIDurationKey,
+			Help:           "Duration in seconds to start the first network pod, excluding time to pull images and run init containers, measured from pod creation timestamp to when all its containers are reported as started and observed via watch",
+			StabilityLevel: metrics.ALPHA,
+		},
+	)
+
+	// FirstNetworkPodStartTotalDuration is a gauge that tracks the duration (in seconds) it takes for the first network pod to run
+	// since creation, including the time for image pulling.
+	FirstNetworkPodStartTotalDuration = metrics.NewGauge(
+		&metrics.GaugeOpts{
+			Subsystem:      KubeletSubsystem,
+			Name:           FirstNetworkPodStartTotalDurationKey,
+			Help:           "Duration in seconds to start the first network pod since creation, including time to pull images and run init containers, measured from pod creation timestamp to when all its containers are reported as started and observed via watch",
+			StabilityLevel: metrics.ALPHA,
+		},
+	)
```
These are very strange metrics.. are you basically just recording this a single time?

aojea replied:
yes, the goal is to measure the user perceived latency when running workloads, we already have node startup latency https://github.com/kubernetes/kubernetes/pull/118568 and pod startup latency, but the latency perceived by an user will  be the "node startup + first pod with network latency"

dgrisonnet replied:
@aojea is this the only place at which we can measure CNI plugin initialization? Because measuring it at the pod startup level kind of pollutes the data with other variables such as the image pulling latency.

logicalhan commented on pkg/kubelet/util/pod_startup_latency_tracker_test.go:
```diff
@@ -275,6 +275,72 @@ kubelet_pod_start_sli_duration_seconds_count 1
+		tracker.DeletePodStartupState(podStarted.UID)
+
+		assert.Empty(t, tracker.pods)
+		metrics.PodStartSLIDuration.Reset()
```
Why are you resetting this metric, you're not even testing this one... shouldn't you be setting the gauge values of the tested metrics to 0?

aojea replied:
the metrics has to be reseted always or when you do some test with `-count N` or with other test with metrics in parallel they fail, since the metrics registry is global, see https://github.com/kubernetes/kubernetes/pull/100218

logicalhan reviewed (commented):
I'm not crazy about these metrics, this kind of data feels much better suited for logs.

aojea commented:
alternative in https://github.com/kubernetes/kubernetes/pull/121721

dgrisonnet commented on pkg/kubelet/metrics/metrics.go:
```diff
@@ -30,46 +30,48 @@ import (
-	WorkingPodCountKey                 = "working_pods"
-	OrphanedRuntimePodTotalKey         = "orphaned_runtime_pods_total"
-	RestartedPodTotalKey               = "restarted_pods_total"
+	FirstNetworkPodStartSLIDurationKey   = "first_network_pod_start_sli_duration_seconds"
+	FirstNetworkPodStartTotalDurationKey = "first_network_pod_start_total_duration_seconds"
```
The current naming is based on an implementation detail so it may be difficult to understand for the end users.
Based on my current understanding of this metric, I would suggest something along those lines:
```
kubelet_node_startup_network_initialization_delay_duration_seconds
```

aojea replied:
it is not really node startup since it depends on the workload to be scheduled to that node, that is the tricky part ... this metric only represent node_startup if the workload is scheduled as soon as the node is ready

aojea replied:
since is internal and temporary I think this name is ok since it really represents the metric

dgrisonnet commented on pkg/kubelet/metrics/metrics.go:
```diff
@@ -211,6 +213,30 @@ var (
+		},
+	)
+
+	// FirstNetworkPodStartTotalDuration is a gauge that tracks the duration (in seconds) it takes for the first network pod to run
+	// since creation, including the time for image pulling.
```
I don't think it is necessary to have both metrics with and without accounting for the image pulling and init containers latency. As far as I understand the context behind this PR, having a metric dedicated strictly to network initialization latency is what we want in order to detect late startup caused by CNI initialization.

dgrisonnet replied:
I think we should only keep the metric that doesn't account for image pulling and init container latency which is the SLI one, because we already have `pod_start_total_duration_seconds` to measure the latency seen by the users. Here what we specifically want to know is whether this latency is caused by the network initialization on the node or not. So the less extraneous latency it accounts for, the better.

aojea replied:
done

dgrisonnet commented on pkg/kubelet/util/pod_startup_latency_tracker.go:
```diff
@@ -112,6 +114,13 @@ func (p *basicPodStartupLatencyTracker) ObservedPodOnWatch(pod *v1.Pod, when tim
+		if !pod.Spec.HostNetwork && !p.firstNetworkPodSeen {
+			metrics.FirstNetworkPodStartSLIDuration.Set(podStartSLOduration)
+			metrics.FirstNetworkPodStartTotalDuration.Set(podStartingDuration.Seconds())
+			p.firstNetworkPodSeen = true
```
Does this need to also account for CNI config modifications and CNI plugin changes?

dgrisonnet replied:
Discussed offline, the CNI is immutable, so we don't have to account for that.

ruiwen-zhao commented on pkg/kubelet/util/pod_startup_latency_tracker.go:
```diff
@@ -112,6 +114,13 @@ func (p *basicPodStartupLatencyTracker) ObservedPodOnWatch(pod *v1.Pod, when tim
 		state.metricRecorded = true
+		// if is the first Pod with network track the start values
+		// these metrics will help to identify problems with the CNI plugin
+		if !pod.Spec.HostNetwork && !p.firstNetworkPodSeen {
```
For my own education, why are we tracking pods with HostNetwork settings? Is it because these pods can only start after CNI initialization is finished, so we are using them as an indirect indicator of CNI init latency? If so why don't we measure the CNI latency directly?

aojea replied:
> why are we tracking pods with HostNetwork settings? 

just the contrary, only the first Pod without HostNetwork set, that gives us the indirect CNI latency

> If so why don't we measure the CNI latency directly?

This is complex and better explained on this doc

https://docs.google.com/document/d/1Ab2XvcS3koXB50RMhsfxdlXNkej5kW6RyxN1F48uqTs/edit

the TL;DR is that we created a chicken an egg problem and we need to deal with the consequences, existing network readiness is not accurate and the different CNI implementations usually are deployed as daemonset, so the CNI is installed as a hostNetwork Pod, and once it is running all the network pods can start running. ... there are some solutions going on to improve this modifying the CNI spec, but those things can take years in be adopted by all the components involved

wojtek-t commented on pkg/kubelet/metrics/metrics.go:
```diff
@@ -211,6 +213,30 @@ var (
+	// excluding the time for image pulling. This metric should reflect the "Pod startup latency SLI" definition and measure the time
+	// for the CNI plugin to be able to provide the network capabilities to the Pod.
+	// ref: https://github.com/kubernetes/community/blob/master/sig-scalability/slos/pod_startup_latency.md
+	FirstNetworkPodStartSLIDuration = metrics.NewGauge(
```
@aojea - I think you were chatting with @logicalhan about it already and I don't know where exactly you converged, but I wanted to throw one more idea and hear why it bad.

What if instead introducing a new metrics for that, we will introduce a label for the already existing PodStartSLIDuration metrics like "kind"? [better name is needed].

Then we will be able to distinguish different types of pods in that, e.g.:
"kind=first-pod-with-network"
for this particular case.

But I think it would be universally useful, because we will certainly need this for other cases, e.g. we want to be able to distinguish stateless and stateful pods because they clearly have a very different performance characteristic. And we would be able to to reuse the same label for it, e.g.:
kind=stateless
kind=stateful

@logicalhan @dgrisonnet - thoughts?

logicalhan replied:
That was actually what I originally proposed. However, given that this metric is a stopgap until we can properly pipe it via Status or something, I'm okay with an isolated metric which is of stability class `Internal` so that it doesn't get output in our auto generated docs and will be removed after a year or so.

dgrisonnet replied:
Being able to distinguish pods could be useful, but it don't think it would make sense to add the label for the current use case as we are essentially trying to measure something different. The fact that we are measuring it via a specific pod latency is an implementation detail, that I don't think we should surface in the general purpose metric.

But as Han mentioned, the general consensus is to introduce a new `Internal` metric as a temporary solution.

aojea replied:
> kind=stateless

Do we have that granularity on the kubelet?

aojea committed:
kubelet: add metric for the first pod with network latency

The first pod with network latency impact user workloads, however,
it is difficuly to understand where is the problem of this latency,
since it depends on the CNI plugin to be ready at the moment of the
pod creation.

Add a new internal metric in the kubelet that allow developers and cluster
administrator to understand the source of the latency problems on
node startups.

kubelet_first_network_pod_start_sli_duration_seconds

Change-Id: I4cdb55b0df72c96a3a65b78ce2aae404c5195006

dashpole commented:
/triage accepted

bart0sh commented:
/priority important-longterm

aojea commented:
@dgrisonnet @logicalhan @wojtek-t kindly reminder, how can I get this to the finish line?

wojtek-t commented:
@logicalhan - friendly ping

dgrisonnet reviewed (approved):
/lgtm

I think the concerns that we had in the past about this metric have been addressed by making it internal.

wojtek-t commented:
/lgtm
/approve
